react/javascript/src/components/gherkin/MDG.tsx

```diff
@@ -0,0 +1,42 @@
+import React from 'react'
+import ReactMarkdown from 'react-markdown'
+import rehypePlugins from '../app/rehypePlugins'
+import GherkinQueryContext from '../../GherkinQueryContext'
+import Step from './Step'
+import styles from '../app/GherkinDocumentList.module.scss'
+
+type IProps = { uri: string; children: any }
+
+const MDG: React.FunctionComponent<IProps> = ({ uri, children }) => {
+  const gherkinQuery = React.useContext(GherkinQueryContext)
+
+  return (
+    <div className={styles.list}>
+      <ReactMarkdown
+        rehypePlugins={rehypePlugins}
+        components={{
+          ul({ children }) {
+            return <ul className={'cucumber-steps'}>{children}</ul>
+          },
+          li({ node, children }) {
+            const line = node.position.start.line
+            const step = gherkinQuery.getStep(uri, line)
+            if (!step) {
+              // Non-Gherkin list item
+              return <li>{children}</li>
+            }
+            return (
+              <li>
+                <Step step={step} renderStepMatchArguments={true} renderMessage={true} />
```

Comment:
This will render the step *as well as its DataTable*. This results in the Markdown table being rendered twice (once by the `<Step>` component, and once by the Markdown renderer (where it doesn't get any styles).

<img width="387" alt="Screenshot 2021-05-19 at 22 39 07" src="https://user-images.githubusercontent.com/1000/118888781-01672500-b8f4-11eb-8d64-9e661013f1d4.png">


In order to avoid this double rendering I think it would be better to lift the `<DataTable>` rendering out of the `<Step>` component and have `<StepList>` render `<DataTable>` underneath the `<Step>`.

In Markdown mode, we'd let Markdown handle the rendering, but we could register custom `td`, `th` and `tr` compoents in `react-markdown`. We should consider extracting `Td`, `Th` and `Tr` components from `DataTable` so they can use the same components (and get the same styles).

Reply:
Oh interesting. I guess we'd have a similar situation for a doc string - a duplicate `<pre>`?

Pulling them out of the `<Step>` would help the Markdown problem, but would also change the order of things so that the data table or doc string would appear after any attachments or errors for the step - I think this might be a bit confusing to read.

Coming at it from the opposite angle, what if we were to, for `table` and `pre`, check if the next non-whitespace line above is a list item, and if so just render nothing?

Reply:
FWIW pushed a commit that applies our data table style to react-markdown's table rendering.

Although looking at it now, it being presented like a Gherkin table but in a Markdown document feels a bit off 😬

Reply:
> ... but would also change the order of things so that the data table or doc string would appear after any attachments or errors for the step

Couldn't we address that by also lifting `<Attachment>` out of `<Step>`?

Reply:
Okay, so `<Step>` just becomes about the title line, the the other bits become other items in the list. Confused about how we'd still be able to maintain the order for Markdown though?

Reply:
Do you mean how we'd be able to display the error messages and attachments (runtime content) in Markdown?

I wonder if we could add some state to the `table` and `li` components and render runtime content directly after the `<Step>` (if there is no DataTable/DocString), otherwise after the DataTable/DocString.

Reply:
Here's the relevant excerpt of [the Markdown doc we're testing with](https://github.com/cucumber/common/blob/main/compatibility-kit/javascript/features/markdown/markdown.md):

```markdown
### Example: DataTable

Here is an example of what it looks like:

* Given this is a typical table:

| name | age |
| ---- | --: |
| Bill |   3 |
| Jane |   6 |
```

If we follow the same pattern as the steps with doc strings, and put the table immediately after the step bullet, like this...

```markdown
### Example: DataTable

Here is an example of what it looks like:

* Given this is a typical table:
  | name | age |
  | ---- | --: |
  | Bill |   3 |
  | Jane |   6 |
```

...the problem disappears, because the Markdown AST has the `table` as a child of the `li`, and we override the entire rendering of the `li` anyway because it's a step.

Would it be reasonable to mandate that in terms of what's valid MDG?

Reply:
Yes I think that’s reasonable. So we’ll only recognise datatables that have 2 or more spaces indentation? Need to check the spec....